File: day17/main.go
package main

import "fmt"

type (
	Vec2 struct {
		x int
		y int
	}
)

func main() {
	//target area: x=25..67, y=-260..-200
	x1, x2, y1, y2 := 25, 67, -260, -200

	maxvy := 0
	{
		maxy := 0
		lower := 256
		upper := 264
		for upper >= lower {
			vy := lower + (upper-lower)/2
			k, ok := simulate(Vec2{0, 0}, Vec2{7, vy}, x1, x2, y1, y2)
			if !ok {
				upper = vy - 1
			} else {
				if k > maxy {
					maxy = k
					maxvy = vy
				}
				lower = vy + 1
			}
		}
		fmt.Println("Part 1:", maxy)
	}
	{
		count := 0
		// 6th triangular number is 21
		for x := 7; x < x2+1; x++ {
			for y := y1 - 1; y < maxvy+1; y++ {
				_, ok := simulate(Vec2{0, 0}, Vec2{x, y}, x1, x2, y1, y2)
				if ok {
					count++
				}
			}
		}
		fmt.Println("Part 2:", count)
	}
}

func simulate(p Vec2, v Vec2, x1, x2, y1, y2 int) (int, bool) {
	maxy := 0
	for {
		p.x += v.x
		p.y += v.y

		if v.x > 0 {
			v.x--
		} else if v.x < 0 {
			v.x++
		}
		v.y--

		if p.y > maxy {
			maxy = p.y
		}

		if inTarget(p, x1, x2, y1, y2) {
			return maxy, true
		}
		if p.y < y1 {
			return 0, false
		}
		if v.x == 0 && (p.x < x1 || p.x > x2) {
			return 0, false
		}
	}
}

func inTarget(p Vec2, x1, x2, y1, y2 int) bool {
	return p.x >= x1 && p.x <= x2 && p.y >= y1 && p.y <= y2
}

File: day17/src/main.rs
#[derive(Clone, Copy)]
struct Vec2(i32, i32);

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let (x1, x2, y1, y2) = (25, 67, -260, -200);
    let mut maxvy = 0;
    {
        let mut maxy = 0;
        let mut lower = 256;
        let mut upper = 265;
        while upper >= lower {
            let vy = lower + (upper - lower) / 2;
            if let Some(k) = simulate(Vec2(0, 0), Vec2(7, vy), x1, x2, y1, y2) {
                if k > maxy {
                    maxy = k;
                    maxvy = vy;
                }
                lower = vy + 1;
            } else {
                upper = vy - 1;
            }
        }
        println!("Part 1: {}", maxy)
    }
    {
        println!(
            "Part 2: {}",
            (7..x2 + 1)
                .into_iter()
                .flat_map(|i| (y1 - 1..maxvy + 1).into_iter().map(move |j| Vec2(i, j)))
                .filter_map(|p| simulate(Vec2(0, 0), p, x1, x2, y1, y2))
                .count()
        );
    }
    Ok(())
}

fn simulate(
    Vec2(mut x, mut y): Vec2,
    Vec2(mut vx, mut vy): Vec2,
    x1: i32,
    x2: i32,
    y1: i32,
    y2: i32,
) -> Option<i32> {
    let mut maxy = 0;
    loop {
        x += vx;
        y += vy;

        if vx > 0 {
            vx -= 1;
        } else if vx < 0 {
            vx += 1;
        }
        vy -= 1;

        if y > maxy {
            maxy = y;
        }

        if in_target(Vec2(x, y), x1, x2, y1, y2) {
            return Some(maxy);
        }
        if y < y1 {
            return None;
        }
        if vx == 0 && (x < x1 || x > x2) {
            return None;
        }
    }
}

fn in_target(Vec2(x, y): Vec2, x1: i32, x2: i32, y1: i32, y2: i32) -> bool {
    x >= x1 && x <= x2 && y >= y1 && y <= y2
}
